mpc: More aggressive ahead of time optimization for string-key type formatter.

pCYSl5EDgo opened the pull request:
# Summary
Current mpc generated code of string-key formatter unnecessarily allocates byte[] array and others.
This pull request reduces memory allocation, improves runtime initialization performance and embeds specialized automaton code for deserialization.

# Current problems

<details><summary>Sample target string-key type code.</summary><div>

```csharp
using MessagePack;

[MessagePackObject(true)]
public class TestType
{
    public int Value { get; set; }
    public int Value0 { get; set; }
    public int Value1 { get; set; }
    public int Value2 { get; set; }

    public int EightSensesFifth0 { get; set; }
    public int EightSensesFifth1 { get; set; }
    public int EightSensesFifth2 { get; set; }
    public int EightSensesFifth3 { get; set; }
    public int EightSensesFifth4 { get; set; }

    public string _0123456_0123456 { get; set; }
}
```
</div></details>

<details><summary>Current generated formmater code.</summary><div>

```csharp
namespace MessagePack.Formatters
{
    using System;
    using System.Buffers;
    using MessagePack;

    public sealed class TestTypeFormatter : global::MessagePack.Formatters.IMessagePackFormatter<global::TestType>
    {


        private readonly global::MessagePack.Internal.AutomataDictionary ____keyMapping;
        private readonly byte[][] ____stringByteKeys;

        public TestTypeFormatter()
        {
            this.____keyMapping = new global::MessagePack.Internal.AutomataDictionary()
            {
                { "Value", 0 },
                { "Value0", 1 },
                { "Value1", 2 },
                { "Value2", 3 },
                { "EightSensesFifth0", 4 },
                { "EightSensesFifth1", 5 },
                { "EightSensesFifth2", 6 },
                { "EightSensesFifth3", 7 },
                { "EightSensesFifth4", 8 },
                { "_0123456_0123456", 9 },
            };

            this.____stringByteKeys = new byte[][]
            {
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("Value"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("Value0"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("Value1"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("Value2"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("EightSensesFifth0"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("EightSensesFifth1"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("EightSensesFifth2"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("EightSensesFifth3"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("EightSensesFifth4"),
                global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes("_0123456_0123456"),
            };
        }

        public void Serialize(ref MessagePackWriter writer, global::TestType value, global::MessagePack.MessagePackSerializerOptions options)
        {
            if (value == null)
            {
                writer.WriteNil();
                return;
            }

            IFormatterResolver formatterResolver = options.Resolver;
            writer.WriteMapHeader(10);
            writer.WriteRaw(this.____stringByteKeys[0]);
            writer.Write(value.Value);
            writer.WriteRaw(this.____stringByteKeys[1]);
            writer.Write(value.Value0);
            writer.WriteRaw(this.____stringByteKeys[2]);
            writer.Write(value.Value1);
            writer.WriteRaw(this.____stringByteKeys[3]);
            writer.Write(value.Value2);
            writer.WriteRaw(this.____stringByteKeys[4]);
            writer.Write(value.EightSensesFifth0);
            writer.WriteRaw(this.____stringByteKeys[5]);
            writer.Write(value.EightSensesFifth1);
            writer.WriteRaw(this.____stringByteKeys[6]);
            writer.Write(value.EightSensesFifth2);
            writer.WriteRaw(this.____stringByteKeys[7]);
            writer.Write(value.EightSensesFifth3);
            writer.WriteRaw(this.____stringByteKeys[8]);
            writer.Write(value.EightSensesFifth4);
            writer.WriteRaw(this.____stringByteKeys[9]);
            formatterResolver.GetFormatterWithVerify<string>().Serialize(ref writer, value._0123456_0123456, options);
        }

        public global::TestType Deserialize(ref MessagePackReader reader, global::MessagePack.MessagePackSerializerOptions options)
        {
            if (reader.TryReadNil())
            {
                return null;
            }

            options.Security.DepthStep(ref reader);
            IFormatterResolver formatterResolver = options.Resolver;
            var length = reader.ReadMapHeader();
            var __Value__ = default(int);
            var __Value0__ = default(int);
            var __Value1__ = default(int);
            var __Value2__ = default(int);
            var __EightSensesFifth0__ = default(int);
            var __EightSensesFifth1__ = default(int);
            var __EightSensesFifth2__ = default(int);
            var __EightSensesFifth3__ = default(int);
            var __EightSensesFifth4__ = default(int);
            var ___0123456_0123456__ = default(string);

            for (int i = 0; i < length; i++)
            {
                ReadOnlySpan<byte> stringKey = global::MessagePack.Internal.CodeGenHelpers.ReadStringSpan(ref reader);
                int key;
                if (!this.____keyMapping.TryGetValue(stringKey, out key))
                {
                    reader.Skip();
                    continue;
                }

                switch (key)
                {
                    case 0:
                        __Value__ = reader.ReadInt32();
                        break;
                    case 1:
                        __Value0__ = reader.ReadInt32();
                        break;
                    case 2:
                        __Value1__ = reader.ReadInt32();
                        break;
                    case 3:
                        __Value2__ = reader.ReadInt32();
                        break;
                    case 4:
                        __EightSensesFifth0__ = reader.ReadInt32();
                        break;
                    case 5:
                        __EightSensesFifth1__ = reader.ReadInt32();
                        break;
                    case 6:
                        __EightSensesFifth2__ = reader.ReadInt32();
                        break;
                    case 7:
                        __EightSensesFifth3__ = reader.ReadInt32();
                        break;
                    case 8:
                        __EightSensesFifth4__ = reader.ReadInt32();
                        break;
                    case 9:
                        ___0123456_0123456__ = formatterResolver.GetFormatterWithVerify<string>().Deserialize(ref reader, options);
                        break;
                    default:
                        reader.Skip();
                        break;
                }
            }

            var ____result = new global::TestType();
            ____result.Value = __Value__;
            ____result.Value0 = __Value0__;
            ____result.Value1 = __Value1__;
            ____result.Value2 = __Value2__;
            ____result.EightSensesFifth0 = __EightSensesFifth0__;
            ____result.EightSensesFifth1 = __EightSensesFifth1__;
            ____result.EightSensesFifth2 = __EightSensesFifth2__;
            ____result.EightSensesFifth3 = __EightSensesFifth3__;
            ____result.EightSensesFifth4 = __EightSensesFifth4__;
            ____result._0123456_0123456 = ___0123456_0123456__;
            reader.Depth--;
            return ____result;
        }
    }
}
```
</div></details>

## ____stringByteKeys
Current generated `TestTypeFormatter` has a byte[][] field `____stringByteKeys`. `____stringByteKeys` does not need to be byte[][]. Accessing `____stringByteKeys` in `Serialize` method results in a performance penalty because of array boundary check.

## No need for runtime call of global::MessagePack.Internal.CodeGenHelpers.GetEncodedStringBytes
Initialization of `____stringByteKeys` uses unnecessary cpu power. String keys are known before runtime execution. Encoded string byte keys can be calculated by mpc and can be embedded as array initialization list.

## ____keyMapping
`____keyMapping` is a `global::MessagePack.Internal.AutomataDictionary` type field. `global::MessagePack.Internal.AutomataDictionary` allocates unnecessary arrays. `AutomataDictionary` builds a trie tree from given string-keys in constructor and performs an automaton search in `Deserialize` method.

# Achieved Pros

<details><summary>Generated formatter code by this pull request.</summary><div>

```csharp
namespace MessagePack.Formatters
{
    using System;
    using System.Buffers;
    using System.Runtime.InteropServices;
    using MessagePack;

    public sealed class TestTypeFormatter : global::MessagePack.Formatters.IMessagePackFormatter<global::TestType>
    {
        private static byte[] ____stringByteKeys0;
        private static byte[] ____stringByteKeys1;
        private static byte[] ____stringByteKeys2;
        private static byte[] ____stringByteKeys3;
        private static byte[] ____stringByteKeys4;
        private static byte[] ____stringByteKeys5;
        private static byte[] ____stringByteKeys6;
        private static byte[] ____stringByteKeys7;
        private static byte[] ____stringByteKeys8;
        private static byte[] ____stringByteKeys9;
        public TestTypeFormatter()
        {
            ____stringByteKeys0 = new byte[]{ 0xA5, 0x56, 0x61, 0x6C, 0x75, 0x65, };
            ____stringByteKeys1 = new byte[]{ 0xA6, 0x56, 0x61, 0x6C, 0x75, 0x65, 0x30, };
            ____stringByteKeys2 = new byte[]{ 0xA6, 0x56, 0x61, 0x6C, 0x75, 0x65, 0x31, };
            ____stringByteKeys3 = new byte[]{ 0xA6, 0x56, 0x61, 0x6C, 0x75, 0x65, 0x32, };
            ____stringByteKeys4 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x30, };
            ____stringByteKeys5 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x31, };
            ____stringByteKeys6 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x32, };
            ____stringByteKeys7 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x33, };
            ____stringByteKeys8 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x34, };
            ____stringByteKeys9 = new byte[]{ 0xB0, 0x5F, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x5F, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, };
        }

        public void Serialize(ref MessagePackWriter writer, global::TestType value, global::MessagePack.MessagePackSerializerOptions options)
        {
            if (value == null)
            {
                writer.WriteNil();
                return;
            }

            IFormatterResolver formatterResolver = options.Resolver;
            writer.WriteMapHeader(10);
            writer.WriteRaw(____stringByteKeys0);
            writer.Write(value.Value);
            writer.WriteRaw(____stringByteKeys1);
            writer.Write(value.Value0);
            writer.WriteRaw(____stringByteKeys2);
            writer.Write(value.Value1);
            writer.WriteRaw(____stringByteKeys3);
            writer.Write(value.Value2);
            writer.WriteRaw(____stringByteKeys4);
            writer.Write(value.EightSensesFifth0);
            writer.WriteRaw(____stringByteKeys5);
            writer.Write(value.EightSensesFifth1);
            writer.WriteRaw(____stringByteKeys6);
            writer.Write(value.EightSensesFifth2);
            writer.WriteRaw(____stringByteKeys7);
            writer.Write(value.EightSensesFifth3);
            writer.WriteRaw(____stringByteKeys8);
            writer.Write(value.EightSensesFifth4);
            writer.WriteRaw(____stringByteKeys9);
            formatterResolver.GetFormatterWithVerify<string>().Serialize(ref writer, value._0123456_0123456, options);
        }

        public global::TestType Deserialize(ref MessagePackReader reader, global::MessagePack.MessagePackSerializerOptions options)
        {
            if (reader.TryReadNil())
            {
                return null;
            }

            options.Security.DepthStep(ref reader);
            IFormatterResolver formatterResolver = options.Resolver;
            var __Value__ = default(int);
            var __Value0__ = default(int);
            var __Value1__ = default(int);
            var __Value2__ = default(int);
            var __EightSensesFifth0__ = default(int);
            var __EightSensesFifth1__ = default(int);
            var __EightSensesFifth2__ = default(int);
            var __EightSensesFifth3__ = default(int);
            var __EightSensesFifth4__ = default(int);
            var ___0123456_0123456__ = default(string);
            var isBigEndian = !global::System.BitConverter.IsLittleEndian;

            for (int i = 0, length = reader.ReadMapHeader(); i < length; i++)
            {
                var stringKey = global::MessagePack.Internal.CodeGenHelpers.ReadStringSpan(ref reader);
                switch(stringKey.Length)
                {
                    case 5:
                        {
                            ulong last = stringKey[4];
                            last <<= 8;
                            last |= stringKey[3];
                            last <<= 8;
                            last |= stringKey[2];
                            last <<= 8;
                            last |= stringKey[1];
                            last <<= 8;
                            last |= stringKey[0];
                            switch (last)
                            {
                                default: goto FAIL;
                                case 0x65756C6156UL:
                                    __Value__ = reader.ReadInt32();
                                    continue;
                            }
                        }
                    case 6:
                        {
                            ulong last = stringKey[5];
                            last <<= 8;
                            last |= stringKey[4];
                            last <<= 8;
                            last |= stringKey[3];
                            last <<= 8;
                            last |= stringKey[2];
                            last <<= 8;
                            last |= stringKey[1];
                            last <<= 8;
                            last |= stringKey[0];
                            switch (last)
                            {
                                default: goto FAIL;
                                case 0x3065756C6156UL:
                                    __Value0__ = reader.ReadInt32();
                                    continue;
                                case 0x3165756C6156UL:
                                    __Value1__ = reader.ReadInt32();
                                    continue;
                                case 0x3265756C6156UL:
                                    __Value2__ = reader.ReadInt32();
                                    continue;
                            }
                        }
                }

                ReadOnlySpan<ulong> ulongs = isBigEndian ? stackalloc ulong[stringKey.Length >> 3] : MemoryMarshal.Cast<byte, ulong>(stringKey);
                if(isBigEndian)
                {
                    for(var index = 0; index < ulongs.Length; index++)
                    {
                        var index8times = index << 3;
                        ref var number = ref global::System.Runtime.CompilerServices.Unsafe.AsRef(ulongs[index]);
                        number = stringKey[index8times + 7];
                        for (var numberIndex = index8times + 6; numberIndex >= index8times; numberIndex--)
                        {
                            number <<= 8;
                            number |= stringKey[numberIndex];
                        }
                    }
                }

                switch(stringKey.Length)
                {
                    default: goto FAIL;
                    case 16:
                        if (ulongs[0] != 0x363534333231305FUL ||
                            ulongs[1] != 0x363534333231305FUL) goto FAIL;

                        ___0123456_0123456__ = formatterResolver.GetFormatterWithVerify<string>().Deserialize(ref reader, options);
                        continue;
                    case 17:
                        if (ulongs[0] != 0x6E65537468676945UL) goto FAIL;
                        if (ulongs[1] != 0x6874666946736573UL) goto FAIL;
                        {
                            uint last = stringKey[16];
                            switch (last)
                            {
                                default: goto FAIL;
                                case 0x30U:
                                    __EightSensesFifth0__ = reader.ReadInt32();
                                    continue;
                                case 0x31U:
                                    __EightSensesFifth1__ = reader.ReadInt32();
                                    continue;
                                case 0x32U:
                                    __EightSensesFifth2__ = reader.ReadInt32();
                                    continue;
                                case 0x33U:
                                    __EightSensesFifth3__ = reader.ReadInt32();
                                    continue;
                                case 0x34U:
                                    __EightSensesFifth4__ = reader.ReadInt32();
                                    continue;
                            }
                        }
                }

            FAIL:
                reader.Skip();
            }

            var ____result = new global::TestType();
            ____result.Value = __Value__;
            ____result.Value0 = __Value0__;
            ____result.Value1 = __Value1__;
            ____result.Value2 = __Value2__;
            ____result.EightSensesFifth0 = __EightSensesFifth0__;
            ____result.EightSensesFifth1 = __EightSensesFifth1__;
            ____result.EightSensesFifth2 = __EightSensesFifth2__;
            ____result.EightSensesFifth3 = __EightSensesFifth3__;
            ____result.EightSensesFifth4 = __EightSensesFifth4__;
            ____result._0123456_0123456 = ___0123456_0123456__;
            reader.Depth--;
            return ____result;
        }
    }
}
```
</div></details>

## Embedding raw byte array initialization list
`____stringByteKeys0 = new byte[]{ 0xA5, 0x56, 0x61, 0x6C, 0x75, 0x65, };` is compiled by C# compiler and IL codes are like [this link](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJmADgE5gBuAhsAKaYBGAnowNoC6mA+v7xgIADACF2DANINWAZ2GYAvJgB2DAO4sJ3ApmEAPAIIBWADR79xgGzmDVhLf1XcjgOxmLVjwF8A3Km8gA==). `System.Runtime.CompilerServices.RuntimeHelpers::InitializeArray` is very performant method.

## Inlined searching of `Deserialize`
Embedded inlined searching codes are now generated. Generated code's literal is Little-Endian. The performance penalty will appear when the runtime endian is Big-Endian.

## Remove line `IFormatterResolver formatterResolver = options.Resolver;` when not needed.
`IFormatterResolver formatterResolver = options.Resolver;` is not needed when there are no field whose type is primitive or has a custom formatter. This pull request removes unnecessary line from generated codes.

AArnott commented:
Wow. I've no doubt this is an improvement. Thanks for contributing!
But I have two questions:

1. How much faster is it really? And I mean for the typical use case not some contrived case that does nothing other than exercise the code that has been changed. 
1. Is the speed improvement worth the substantially decreased readability of the generated code? I was just recommending earlier this week that someone who needed customized serialization to use mpc to generate their own formatter/resolver and then change the output to meet their needs. They can do this today, but after this PR, I can't imagine *anyone* would want to use mpc output as a basis for customizing their formatter. If the perf improvement is substantial, and most people don't use mpc to customize serialization, then perhaps the loss in readability is worth it.

@neuecc, what do you think?

neuecc commented:
There are a number of improvements in there, but I think the basic point is that mpc also does the inlined constant embedding automata that we do during dynamic code generation.
When I first introduced to automaton, I omit inlining to mpc.
http://neue.cc/2017/08/28_558.html
However, I agree with this PR policy because I think it's better to be close.

I've also made custom formatters based on code generation by MPC and modified them.
That said, the current MPC-generated results are also by no means easy to fix, so you may not have to think about that.
However, replacing GetEncodedStringBytes with a new byte is a complete loss of name, so I think then __stringByteKeysX needs to keep the name information, like __EightSensesFifth0Key__.

Also, I'd like the code to be similar to the current dynamic code generation.
For example, using AutomataKeyGen.GetKey to use the same key, and so on.
This is for maintenance purposes.
I don't have the confidence to do the maintenance if it generates code that is clearly of a different strain.

pCYSl5EDgo commented:
> However, replacing GetEncodedStringBytes with a new byte is a complete loss of name, so I think then __stringByteKeysX needs to keep the name information, like __EightSensesFifth0Key__.
I agree with this comment. I changed the naming rule from `____stringByteKey(Number)` to `____stringByteKey_(Member.Name)`.

> decreased readability
The main cause of readability decrease is the little-endian ulong keys embedded in switch case or if statements. I added comment lines for each of the ulong keys which contain only ascii compatible utf8 characters.

<details><summary>Current Generated Formatter Code with 2 additional commits.</summary><div>

```csharp
namespace MessagePack.Formatters
{
    using System;
    using System.Buffers;
    using System.Runtime.InteropServices;
    using MessagePack;

    public sealed class TestTypeFormatter : global::MessagePack.Formatters.IMessagePackFormatter<global::TestType>
    {
        private static byte[] ____stringByteKeys_Value;
        private static byte[] ____stringByteKeys_Value0;
        private static byte[] ____stringByteKeys_Value1;
        private static byte[] ____stringByteKeys_Value2;
        private static byte[] ____stringByteKeys_EightSensesFifth0;
        private static byte[] ____stringByteKeys_EightSensesFifth1;
        private static byte[] ____stringByteKeys_EightSensesFifth2;
        private static byte[] ____stringByteKeys_EightSensesFifth3;
        private static byte[] ____stringByteKeys_EightSensesFifth4;
        private static byte[] ____stringByteKeys__0123456_0123456;

        public TestTypeFormatter()
        {
            ____stringByteKeys_Value = new byte[]{ 0xA5, 0x56, 0x61, 0x6C, 0x75, 0x65, };
            ____stringByteKeys_Value0 = new byte[]{ 0xA6, 0x56, 0x61, 0x6C, 0x75, 0x65, 0x30, };
            ____stringByteKeys_Value1 = new byte[]{ 0xA6, 0x56, 0x61, 0x6C, 0x75, 0x65, 0x31, };
            ____stringByteKeys_Value2 = new byte[]{ 0xA6, 0x56, 0x61, 0x6C, 0x75, 0x65, 0x32, };
            ____stringByteKeys_EightSensesFifth0 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x30, };
            ____stringByteKeys_EightSensesFifth1 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x31, };
            ____stringByteKeys_EightSensesFifth2 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x32, };
            ____stringByteKeys_EightSensesFifth3 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x33, };
            ____stringByteKeys_EightSensesFifth4 = new byte[]{ 0xB1, 0x45, 0x69, 0x67, 0x68, 0x74, 0x53, 0x65, 0x6E, 0x73, 0x65, 0x73, 0x46, 0x69, 0x66, 0x74, 0x68, 0x34, };
            ____stringByteKeys__0123456_0123456 = new byte[]{ 0xB0, 0x5F, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x5F, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, };
        }

        public void Serialize(ref MessagePackWriter writer, global::TestType value, global::MessagePack.MessagePackSerializerOptions options)
        {
            if (value == null)
            {
                writer.WriteNil();
                return;
            }

            IFormatterResolver formatterResolver = options.Resolver;
            writer.WriteMapHeader(10);
            writer.WriteRaw(____stringByteKeys_Value);
            writer.Write(value.Value);
            writer.WriteRaw(____stringByteKeys_Value0);
            writer.Write(value.Value0);
            writer.WriteRaw(____stringByteKeys_Value1);
            writer.Write(value.Value1);
            writer.WriteRaw(____stringByteKeys_Value2);
            writer.Write(value.Value2);
            writer.WriteRaw(____stringByteKeys_EightSensesFifth0);
            writer.Write(value.EightSensesFifth0);
            writer.WriteRaw(____stringByteKeys_EightSensesFifth1);
            writer.Write(value.EightSensesFifth1);
            writer.WriteRaw(____stringByteKeys_EightSensesFifth2);
            writer.Write(value.EightSensesFifth2);
            writer.WriteRaw(____stringByteKeys_EightSensesFifth3);
            writer.Write(value.EightSensesFifth3);
            writer.WriteRaw(____stringByteKeys_EightSensesFifth4);
            writer.Write(value.EightSensesFifth4);
            writer.WriteRaw(____stringByteKeys__0123456_0123456);
            formatterResolver.GetFormatterWithVerify<string>().Serialize(ref writer, value._0123456_0123456, options);
        }

        public global::TestType Deserialize(ref MessagePackReader reader, global::MessagePack.MessagePackSerializerOptions options)
        {
            if (reader.TryReadNil())
            {
                return null;
            }

            options.Security.DepthStep(ref reader);
            IFormatterResolver formatterResolver = options.Resolver;
            var __Value__ = default(int);
            var __Value0__ = default(int);
            var __Value1__ = default(int);
            var __Value2__ = default(int);
            var __EightSensesFifth0__ = default(int);
            var __EightSensesFifth1__ = default(int);
            var __EightSensesFifth2__ = default(int);
            var __EightSensesFifth3__ = default(int);
            var __EightSensesFifth4__ = default(int);
            var ___0123456_0123456__ = default(string);
            var isBigEndian = !global::System.BitConverter.IsLittleEndian;

            for (int i = 0, length = reader.ReadMapHeader(); i < length; i++)
            {
                var stringKey = global::MessagePack.Internal.CodeGenHelpers.ReadStringSpan(ref reader);
                switch(stringKey.Length)
                {
                    case 5:
                        {
                            ulong last = stringKey[4];
                            last <<= 8;
                            last |= stringKey[3];
                            last <<= 8;
                            last |= stringKey[2];
                            last <<= 8;
                            last |= stringKey[1];
                            last <<= 8;
                            last |= stringKey[0];
                            switch (last)
                            {
                                default: goto FAIL;
                                case 0x65756C6156UL:
                                    __Value__ = reader.ReadInt32();
                                    continue;
                            }
                        }
                    case 6:
                        {
                            ulong last = stringKey[5];
                            last <<= 8;
                            last |= stringKey[4];
                            last <<= 8;
                            last |= stringKey[3];
                            last <<= 8;
                            last |= stringKey[2];
                            last <<= 8;
                            last |= stringKey[1];
                            last <<= 8;
                            last |= stringKey[0];
                            switch (last)
                            {
                                default: goto FAIL;
                                case 0x3065756C6156UL:
                                    __Value0__ = reader.ReadInt32();
                                    continue;
                                case 0x3165756C6156UL:
                                    __Value1__ = reader.ReadInt32();
                                    continue;
                                case 0x3265756C6156UL:
                                    __Value2__ = reader.ReadInt32();
                                    continue;
                            }
                        }
                }

                ReadOnlySpan<ulong> ulongs = isBigEndian ? stackalloc ulong[stringKey.Length >> 3] : MemoryMarshal.Cast<byte, ulong>(stringKey);
                if(isBigEndian)
                {
                    for(var index = 0; index < ulongs.Length; index++)
                    {
                        var index8times = index << 3;
                        ref var number = ref global::System.Runtime.CompilerServices.Unsafe.AsRef(ulongs[index]);
                        number = stringKey[index8times + 7];
                        for (var numberIndex = index8times + 6; numberIndex >= index8times; numberIndex--)
                        {
                            number <<= 8;
                            number |= stringKey[numberIndex];
                        }
                    }
                }

                switch(stringKey.Length)
                {
                    default: goto FAIL;
                    case 16:
                        if (
                            ulongs[0] != 0x363534333231305FUL || // _0123456
                            ulongs[1] != 0x363534333231305FUL // _0123456
                        ) goto FAIL;

                        ___0123456_0123456__ = formatterResolver.GetFormatterWithVerify<string>().Deserialize(ref reader, options);
                        continue;
                    case 17:
                        if (ulongs[0] != 0x6E65537468676945UL) goto FAIL; // EightSen
                        if (ulongs[1] != 0x6874666946736573UL) goto FAIL; // sesFifth
                        {
                            uint last = stringKey[16];
                            switch (last)
                            {
                                default: goto FAIL;
                                case 0x30U:
                                    __EightSensesFifth0__ = reader.ReadInt32();
                                    continue;
                                case 0x31U:
                                    __EightSensesFifth1__ = reader.ReadInt32();
                                    continue;
                                case 0x32U:
                                    __EightSensesFifth2__ = reader.ReadInt32();
                                    continue;
                                case 0x33U:
                                    __EightSensesFifth3__ = reader.ReadInt32();
                                    continue;
                                case 0x34U:
                                    __EightSensesFifth4__ = reader.ReadInt32();
                                    continue;
                            }
                        }
                }

            FAIL:
                reader.Skip();
            }

            var ____result = new global::TestType();
            ____result.Value = __Value__;
            ____result.Value0 = __Value0__;
            ____result.Value1 = __Value1__;
            ____result.Value2 = __Value2__;
            ____result.EightSensesFifth0 = __EightSensesFifth0__;
            ____result.EightSensesFifth1 = __EightSensesFifth1__;
            ____result.EightSensesFifth2 = __EightSensesFifth2__;
            ____result.EightSensesFifth3 = __EightSensesFifth3__;
            ____result.EightSensesFifth4 = __EightSensesFifth4__;
            ____result._0123456_0123456 = ___0123456_0123456__;
            reader.Depth--;
            return ____result;
        }
    }
}
```
</div></details>

pCYSl5EDgo commented:
I made a benchmark project to know how much my pull request is faster than original one.
[Gist](https://gist.github.com/pCYSl5EDgo/71e16324b5ad44393bb0c1f2f1999714).
Deserialization performance is approximately 30% up.

``` ini

BenchmarkDotNet=v0.12.0, OS=Windows 10.0.18363
Intel Core i7-8750H CPU 2.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.1.200
  [Host]     : .NET Core 3.1.2 (CoreCLR 4.700.20.6602, CoreFX 4.700.20.6702), X64 RyuJIT
  Job-ITDJDY : .NET Core 3.1.2 (CoreCLR 4.700.20.6602, CoreFX 4.700.20.6702), X64 RyuJIT

Runtime=.NET Core 3.1  

```
|                                      Method |       ConstantString |     Mean |    Error |   StdDev |
|-------------------------------------------- |--------------------- |---------:|---------:|---------:|
|        **DeserializeTest_NewEmbeddedFormatter** |                     **** | **508.6 ns** | **10.20 ns** | **10.47 ns** |
| DeserializeTest2_OriginalGeneratedFormatter |                      | 682.1 ns | 13.51 ns | 29.66 ns |
|        **DeserializeTest_NewEmbeddedFormatter** |                    **a** | **489.3 ns** |  **6.36 ns** |  **5.95 ns** |
| DeserializeTest2_OriginalGeneratedFormatter |                    a | 694.0 ns |  7.68 ns |  6.81 ns |
|        **DeserializeTest_NewEmbeddedFormatter** |              **shorter** | **537.2 ns** |  **2.44 ns** |  **2.16 ns** |
| DeserializeTest2_OriginalGeneratedFormatter |              shorter | 714.6 ns | 10.28 ns |  8.58 ns |
|        **DeserializeTest_NewEmbeddedFormatter** | **このベンチ(...)おおおん！ [62]** | **675.5 ns** |  **6.43 ns** |  **6.02 ns** |
| DeserializeTest2_OriginalGeneratedFormatter | このベンチ(...)おおおん！ [62] | 871.8 ns |  9.12 ns |  7.12 ns |

waitxd commented on src/MessagePack.GeneratorCore/Generator/StringKey/StringKeyFormatterTemplate.tt:
```diff
@@ -0,0 +1,284 @@
+
+    if (isFormatterResolverNecessary)
+    {#>
+            IFormatterResolver formatterResolver = options.Resolver;
```
```suggestion
            var formatterResolver = options.Resolver;
```

AArnott replied:
I prefer explicit types when the type isn't obvious (because it appears elsewhere on the line). Also: it reduces the effective diff in the generated.cs file for the PR. We're going for functional improvements in this PR -- not style changes.

waitxd commented on src/MessagePack.GeneratorCore/Generator/StringKey/StringKeyFormatterTemplate.tt:
```diff
@@ -0,0 +1,284 @@
+    if (isFormatterResolverNecessary)
+    {
+#>
+            IFormatterResolver formatterResolver = options.Resolver;
```
```suggestion
            var formatterResolver = options.Resolver;
```

waitxd commented on src/MessagePack.GeneratorCore/Generator/FormatterTemplate.cs:
```diff
@@ -62,70 +63,35 @@ namespace ");
+ }
+
+  if (isFormatterResolverNecessary) { 
             this.Write("            IFormatterResolver formatterResolver = options.Resolver;\r\n");
```
```suggestion
            this.Write("            var formatterResolver = options.Resolver;\r\n");
```

waitxd commented on src/MessagePack.GeneratorCore/Generator/FormatterTemplate.cs:
```diff
@@ -138,34 +104,20 @@ namespace ");
-            this.Write("            var length = reader.ReadArrayHeader();\r\n");
+            this.Write("            }\r\n\r\n            options.Security.DepthStep(ref reader);\r\n");
+ if (isFormatterResolverNecessary) { 
+            this.Write("            IFormatterResolver formatterResolver = options.Resolver;\r\n");
```
```suggestion
            this.Write("            var formatterResolver = options.Resolver;\r\n");
```

waitxd commented on src/MessagePack.GeneratorCore/Generator/FormatterTemplate.tt:
```diff
@@ -66,28 +44,23 @@ namespace <#= Namespace #>
+<# }
+
+  if (isFormatterResolverNecessary) { #>
             IFormatterResolver formatterResolver = options.Resolver;
```
```suggestion
            var formatterResolver = options.Resolver;
```

waitxd commented on src/MessagePack.GeneratorCore/Generator/FormatterTemplate.tt:
```diff
@@ -103,31 +76,17 @@ namespace <#= Namespace #>
 
             options.Security.DepthStep(ref reader);
+<# if (isFormatterResolverNecessary) { #>
             IFormatterResolver formatterResolver = options.Resolver;
```
```suggestion
            var formatterResolver = options.Resolver;
```

waitxd commented on src/MessagePack.GeneratorCore/Generator/StringKey/StringKeyFormatterTemplate.tt:
```diff
@@ -0,0 +1,284 @@
+<#
+    }#>
+
+        public <#= formatterName #>()
```
You are using constructor to initialize static fields. Either use static constructor or move static this code into field initialization.

waitxd commented on src/MessagePack.GeneratorCore/Generator/StringKey/StringKeyFormatterTemplate.tt:
```diff
@@ -0,0 +1,284 @@
+                    }
+                }
+
+                switch(stringKey.Length)
```
```suggestion
                switch (stringKey.Length)
```

waitxd commented on src/MessagePack.GeneratorCore/Generator/StringKey/StringKeyFormatterTemplate.tt:
```diff
@@ -0,0 +1,284 @@
+        if(greaterThan7.Length == 0)
+        {#>
+                var stringKey = global::MessagePack.Internal.CodeGenHelpers.ReadStringSpan(ref reader);
+                switch(stringKey.Length)
```
```suggestion
                switch (stringKey.Length)
```

waitxd commented on src/MessagePack.GeneratorCore/Generator/StringKey/StringKeyFormatterTemplate.tt:
```diff
@@ -0,0 +1,284 @@
+        else
+        {#>
+                var stringKey = global::MessagePack.Internal.CodeGenHelpers.ReadStringSpan(ref reader);
+                switch(stringKey.Length)
```
```suggestion
                switch (stringKey.Length)
```

waitxd commented on src/MessagePack.GeneratorCore/Generator/StringKey/StringKeyFormatterTemplate.tt:
```diff
@@ -0,0 +1,284 @@
+                    }
+                }
+
+                switch(stringKey.Length)
```
```suggestion
                switch (stringKey.Length)
```

pCYSl5EDgo commented:
Thank you @waitxd !

> You are using constructor to initialize static fields. Either use static constructor or move static this code into field initialization.
You are right.
I would like to embed those byte arrays as writer.WriteRaw's ReadOnlySpan&lt;byte&gt; parameter.
It is because C# 7.3 and C# compiler remove manage heap allocation.

waitxd commented:
Found two more bugs:
1. If `mpc` output to directory instead of `.cs` file - your template is not used at all.
2. You are checking `MessagePackObject` attribute value only for the first class from each namespace to decide whether should you apply your formatter.

In this example both will be serialized as an array:
```csharp
namespace SharedData
{
    [MessagePackObject()]
    public class SerializeMeAsArrayPlease
    {
        [Key(0)]
        public bool Foo;
    }

    [MessagePackObject(true)]
    public class TotallyDoNotSerializeMeAsArrayPlease
    {
        public bool Foo;
    }
}
```

pCYSl5EDgo commented:
> Bug report

Thank you so much!

> Maybe we should add a comment with meaning of inlined byte arrays in Serialize method as it was done in Deserialize?

It should be embedded. I am worried that the end-user programmer at Unity environment uses these generated codes in .NET 3.5 API Level and version of C# is less than 7.3.

waitxd commented:
You are still generate constructor and static fields that are not used anymore.
<details>
  <summary>Example</summary>

```csharp
    public sealed class StringKeySerializerTargetFormatter : global::MessagePack.Formatters.IMessagePackFormatter<global::PerfBenchmarkDotNet.StringKeySerializerTarget>
    {
        private static byte[] ____stringByteKeys_MyProperty1;

        static StringKeySerializerTargetFormatter()
        {
            ____stringByteKeys_MyProperty1 = new byte[] { 0xAB, 0x4D, 0x79, 0x50, 0x72, 0x6F, 0x70, 0x65, 0x72, 0x74, 0x79, 0x31, };
        }

        public void Serialize(ref MessagePackWriter writer, global::PerfBenchmarkDotNet.StringKeySerializerTarget value, global::MessagePack.MessagePackSerializerOptions options)
        {
            if (value == null)
            {
                writer.WriteNil();
                return;
            }

            writer.WriteMapHeader(1);
            writer.WriteRaw(new byte[] { 0xAB, 0x4D, 0x79, 0x50, 0x72, 0x6F, 0x70, 0x65, 0x72, 0x74, 0x79, 0x31, });
            writer.Write(value.MyProperty1);
        }

        public global::PerfBenchmarkDotNet.StringKeySerializerTarget Deserialize(ref MessagePackReader reader, global::MessagePack.MessagePackSerializerOptions options)
        {
            if (reader.TryReadNil())
            {
                return null;
            }

            options.Security.DepthStep(ref reader);
            var __MyProperty1__ = default(int);
            var isBigEndian = !global::System.BitConverter.IsLittleEndian;

            for (int i = 0, length = reader.ReadMapHeader(); i < length; i++)
            {
                var stringKey = global::MessagePack.Internal.CodeGenHelpers.ReadStringSpan(ref reader);
                ReadOnlySpan<ulong> ulongs = isBigEndian ? stackalloc ulong[stringKey.Length >> 3] : MemoryMarshal.Cast<byte, ulong>(stringKey);
                if (isBigEndian)
                {
                    for (var index = 0; index < ulongs.Length; index++)
                    {
                        var index8times = index << 3;
                        ref var number = ref global::System.Runtime.CompilerServices.Unsafe.AsRef(ulongs[index]);
                        number = stringKey[index8times + 7];
                        for (var numberIndex = index8times + 6; numberIndex >= index8times; numberIndex--)
                        {
                            number <<= 8;
                            number |= stringKey[numberIndex];
                        }
                    }
                }

                switch (stringKey.Length)
                {
                    default:
                    FAIL:
                        reader.Skip();
                        continue;
                    case 11:
                        if (
                            ulongs[0] != 0x7265706F7250794DUL // MyProper
                        ) goto FAIL;
                        {
                            uint last = stringKey[10];
                            last <<= 8;
                            last |= stringKey[9];
                            last <<= 8;
                            last |= stringKey[9];
                            if(last != 0x317974U) goto FAIL;
                            __MyProperty1__ = reader.ReadInt32();
                            continue;
                        }
                }
            }

            var ____result = new global::PerfBenchmarkDotNet.StringKeySerializerTarget();
            ____result.MyProperty1 = __MyProperty1__;
            reader.Depth--;
            return ____result;
        }
    }

```

</details>



> It should be embedded.

Yes, But for those who will read generated code byte array meaning will be unclear.
We can put a comment right after byte array:
```csharp
writer.WriteMapHeader(9);
writer.WriteRaw(new byte[] { 0xAB, 0x4D, 0x79, 0x50, 0x72, 0x6F, 0x70, 0x65, 0x72, 0x74, 0x79, 0x31, }); //MyProperty1
writer.Write(value.MyProperty1);
```
Or before:
```csharp
// MyProperty2
writer.WriteRaw(new byte[] { 0xAB, 0x4D, 0x79, 0x50, 0x72, 0x6F, 0x70, 0x65, 0x72, 0x74, 0x79, 0x32, });
writer.Write(value.MyProperty2);
```



> I am worried that the end-user programmer at Unity environment uses these generated codes in .NET 3.5 API Level and his/her C# version is less than 7.3.

That's should not be a problem. MessagePack-CSharp supports only Unity 2018.3+. It is mentioned in the [readme](https://github.com/neuecc/MessagePack-CSharp#installation).

pCYSl5EDgo commented:
Thank you for all of your great advices!

> MessagePack-CSharp supports only Unity 2018.3+. It is mentioned in the readme.

Oh! I was releaved to hear that!

pCYSl5EDgo commented:
My proposed automaton first branches by string length. The current string search automaton does not consider the length of the string, and compares it in 8-byte units.
I will write a benchmark code and examine how much this difference affects performance.

AArnott commented:
@neuecc How are you feeling about this change at this point? Is it close enough to the dynamic resolver that you're comfortable maintaining it?

neuecc commented:
Serialize is good.
However, as for Deserialize, as I mentioned before, I don't see why you can't use the `AutomataKeyGen.GetKey` used there.
DynamicObjectResolver's generated code and mpc generated results are so different, that can not acceptable.
Also, in the main scenario, Unity IL2CPP, the behavior on Android (which has been an issue many times, right?) I'm concerned about the problems.

pCYSl5EDgo commented:
Okay, I start rewriting the Deserialization method.

pCYSl5EDgo commented:
Seems done.

AArnott commented on src/MessagePack.GeneratorCore/Generator/ShouldUseFormatterResolverHelper.cs:
```diff
@@ -0,0 +1,43 @@
+    {
+        private static readonly string[] PrimitiveTypes =
+        {
+            "short",
+            "int",
+            "long",
+            "ushort",
+            "uint",
+            "ulong",
+            "float",
+            "double",
+            "bool",
+            "byte",
+            "sbyte",
+            "char",
```
There's a similar list somewhere in the MessagePack library with the same purpose: should it use a resolver to get a formatter for it or treat it as a primitive. It would be nice to share the list if we can find it.

AArnott replied:
I found the list and consolidated this. `byte[]` was missing on this list so I added it and fixed up the codegen to handle it properly.

pCYSl5EDgo replied:
Thank you!

AArnott commented on src/MessagePack.GeneratorCore/Generator/StringKey/EmbedStringHelper.cs:
```diff
@@ -0,0 +1,87 @@
+            var headerLength = GetHeaderLength(binary.Length);
+            Span<byte> header = stackalloc byte[headerLength];
+            EmbedHeader(binary.Length, header);
+            var buffer = new StringBuilder().Append("new byte[").Append(headerLength).Append(" + ").Append(binary.Length).Append("] { ").Append(header[0]);
```
This leads to generated code changing to a less efficient form:

```diff
-writer.WriteRaw(this.____stringByteKeys[1]);
+writer.WriteRaw(new byte[1 + 1] { 161, 97 });
```

This means instead of reusing a single array every time serialization happens, we allocate arrays every time. This seems like a step backwards.

pCYSl5EDgo replied:
Thank you for your review!
Oh, I've missed the cases of the short length array.
[I examined those cases in sharplab.](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHAZgAJ9KBhSgb1MpeqpwBZKBZACgEpGzViIBKcCABMA8gDsMATwDKABwiyAPACEF8AHyUAHpQC8lWXADulYLrgBtALpCSIt6wIA2AkkoBOAHZfYXcAXwBuELcAdQQ0eFEIS15DfkjXd0pxKTlFVXVtOwMFU3MrGzsnF0y3H0o8X3IkKJEIltZY+LhE5IU09pZsmXllNU0dfUoAL1KLa1t4KqYMmpY6hsomyg5glbc2vZFOhKTeKf690KioimouY+7TodzRgom4A1So5ZqAMwB7BASADGAAtKLwAG4QBCUEpoWRGfgDaqrFg4Ah+Xh9dI1K6XUihIA==)

MessagePackWriter.WriteRaw has 2 overloads(ReadOnlySpan<byte> and ReadOnlySequence<byte>).
No array allocation seems to occur.

AArnott replied:
I'm thinking about https://github.com/neuecc/MessagePack-CSharp/pull/1044#issuecomment-693029665 and the fact that mpc output is often used in Unity. Are we sure unity's compiler and il2cpp, etc. will provide adequate optimizations? As you left it in that PR, it sounds like you were not sure the optimizations will be there.

pCYSl5EDgo replied:
I've examined Unity IL2CPP code for the static ReadOnlySpan&lt;byte&gt; optimization.

<details><summary>Unity C# source code</summary>

```csharp
using System;
using System.Globalization;
using TMPro;
using UnityEngine;

public class ArrayEmbed : MonoBehaviour
{
    [SerializeField] private TMP_Text text = default;
    
    void Start()
    {
        text.text = EmbedderHelper.Sum(EmbedderHelper.GetTemp()).ToString(CultureInfo.InvariantCulture);
    }
}


public static class EmbedderHelper
{
    public static ReadOnlySpan<byte> GetTemp() => new byte[1 + 1] { 161 , 12 };

    public static uint Sum(ReadOnlySpan<byte> span)
    {
        if (span.IsEmpty)
        {
            throw new ArgumentOutOfRangeException();
        }

        uint answer = span[0];
        foreach (var b in span.Slice(1))
        {
            answer += b;
        }

        return answer;
    }
}
```
</details>

<details><summary>The most important codes I extract.</summary>

```cpp
// System.ReadOnlySpan`1<System.Byte> EmbedderHelper::GetTemp()
extern "C" IL2CPP_METHOD_ATTR ReadOnlySpan_1_t11883B9D8CFA411A5B2BC1C72CCEB8215DFBE5E6  EmbedderHelper_GetTemp_m0A9B4121AE7DA48DDF6FA423281642A01BA2F505 (const RuntimeMethod* method)
{
	static bool s_Il2CppMethodInitialized;
	if (!s_Il2CppMethodInitialized)
	{
		il2cpp_codegen_initialize_method (EmbedderHelper_GetTemp_m0A9B4121AE7DA48DDF6FA423281642A01BA2F505_MetadataUsageId);
		s_Il2CppMethodInitialized = true;
	}
	{
		// public static ReadOnlySpan<byte> GetTemp() => new byte[1 + 1] { 161 , 12 };
		ReadOnlySpan_1_t11883B9D8CFA411A5B2BC1C72CCEB8215DFBE5E6  L_0;
		memset(&L_0, 0, sizeof(L_0));
		ReadOnlySpan_1__ctor_m7D8EDA9D5C232F7B9E48494FAEC1CD57865932E2((&L_0), (void*)(void*)(il2cpp_codegen_get_field_data(U3CPrivateImplementationDetailsU3E_t4846F3E0DC61BBD6AE249FF1EAAC4CB64097DE4A____42BD4AA0A1EA24FF3EB09F993F11256F6F424768_0_FieldInfo_var)), 2, /*hidden argument*/ReadOnlySpan_1__ctor_m7D8EDA9D5C232F7B9E48494FAEC1CD57865932E2_RuntimeMethod_var);
		return L_0;
	}
}
```

```cpp
inline const char* il2cpp_codegen_get_field_data(RuntimeField* field)
{
    return il2cpp::vm::Field::GetData(field);
}
```

```cpp
const char* Field::GetData(FieldInfo *field)
{
    if (field->type->attrs & FIELD_ATTRIBUTE_HAS_DEFAULT)
    {
        const Il2CppType* type = NULL;
        return Class::GetFieldDefaultValue(field, &type);
    }
    else if (field->type->attrs & FIELD_ATTRIBUTE_HAS_FIELD_RVA)
    {
        IL2CPP_NOT_IMPLEMENTED_NO_ASSERT(Field::GetData, "This works for array initialization data. Revisit any other RVA use case.");
        const Il2CppType* type = NULL;
        return Class::GetFieldDefaultValue(field, &type);
    }
    else
    {
        return NULL;
    }
}
```

```cpp
const char* Class::GetFieldDefaultValue(const FieldInfo *field, const Il2CppType** type)
{
    const Il2CppFieldDefaultValue *entry = MetadataCache::GetFieldDefaultValueForField(field);
    if (entry != NULL)
    {
        *type = MetadataCache::GetIl2CppTypeFromIndex(entry->typeIndex);
        if (entry->dataIndex == kDefaultValueIndexNull)
            return NULL;

        return (const char*)MetadataCache::GetFieldDefaultValueDataFromIndex(entry->dataIndex);
    }
    return NULL;
}
```
</details>

AArnott reviewed (changes requested):
This looks pretty good. One inefficiency should be corrected as noted. @neuecc are you satisfied?

neuecc commented:
@pCYSl5EDgo 
Thanks, great contribution.
I think it is okay.
A few code formatting issues.

* StringKey generates deserialize -> serialize but others is serialize -> deserialize, please change order.
* StringKeyFormatterTemplate.tt's format is slightly agry, are you use auto formatting? All of the automatic formatting on T4 is very poor, so I think it's better to put it on manually.
* I don't checked generated code of zero members but Skip twice? Is okay? https://github.com/neuecc/MessagePack-CSharp/blob/f8311a1c764db277a950ede0067e1922c391c2a2/src/MessagePack.GeneratorCore/Generator/StringKey/StringKeyFormatterTemplate.tt#L92-L93
* Is this okay to support different endian?

pCYSl5EDgo commented:
> I don't checked generated code of zero members but Skip twice? Is okay?

It was okay.
But I rewrite it to be more easily understood.

> Different Endian

This PR uses binary sequence compare and AutomataKeyGen.GetKey.
It is okay.
